### DISCUSSION TITLE
Matching search params object/map/array

### DISCUSSION THREAD

**Discussion by OliverJAsh** [whatwg/urlpattern#60] [Ideas]
I'm curious whether we could match the search params using a richer structure than just `string` so that we don't need to worry about matching characters such as `&` or `=`. For example:

```diff
 const p = new URLPattern({
   protocol: 'https',
   username: '',
   password: '',
   hostname: 'example.com',
   port: '',
   pathname: '/foo/:image.jpg',
-  search: '',
+  searchParams: [['foo', 'bar'], ['abc', '*']],
   hash: '*',
 });
```

For example, this would match a `search` of:

- `foo=bar&abc=123`
- `abc=456&foo=bar`

I thought about whether we could re-use `URLSearchParams` but I'm not sure we can, because we don't want to encode characters for pattern matching. Perhaps we could introduce a `URLSearchParamsPattern`. This would allow it to be used on its own.

**Comment by wanderview** [comment]
I agree that pattern syntax is not really fit for extracting search params since they can be re-ordered.  My current expectation is that code that wants to get parameter values out would need to use `URLSearchParams` to parse the query params and then operate on them individually.

We could consider adding some kind of map-like matching construct like you suggest in the future.  It would be helpful to see use cases where it provides significant benefit over using just `URLSearchParams` itself.
